Index: musl-obstack-2023-02-28-f4385255/configure.ac
===================================================================
--- musl-obstack-2023-02-28-f4385255.orig/configure.ac
+++ musl-obstack-2023-02-28-f4385255/configure.ac
@@ -4,7 +4,7 @@ AC_PREREQ(2.69)
 
 AC_INIT([obstack], [1.2], [pullmoll@t-online.de])
 
-AM_INIT_AUTOMAKE([1.16])
+AM_INIT_AUTOMAKE([1.15])
 AC_CONFIG_MACRO_DIR([m4])
 AC_CONFIG_HEADERS([config.h])
 
