From c3f31c0671f6c84f5e02ec710385e1e8f2936066 Mon Sep 17 00:00:00 2001
From: Peter Hunt <pehunt@redhat.com>
Date: Tue, 2 Feb 2021 16:52:06 -0500
Subject: [PATCH] bump to v2.0.27-dev

Signed-off-by: Peter Hunt <pehunt@redhat.com>
---
 VERSION | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/VERSION b/VERSION
index f6ee962..730b464 100644
--- a/VERSION
+++ b/VERSION
@@ -1 +1 @@
-2.0.26
+2.0.27-dev
